x/slashing/abci.go

```diff
@@ -56,6 +56,55 @@ func BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock, k keeper.Keeper)
 		if writeInfo == nil {
 			panic("Expected slashing write info to be non-nil")
 		}
+		// Check if we need to resize the array if there was recently a change in slashing window size
+		window := k.SignedBlocksWindow(ctx)
+		missedInfo, found := k.GetValidatorMissedBlocks(ctx, writeInfo.ConsAddr)
+		missedBlockLen := int64(len(missedInfo.MissedBlocks))
+		if found && window != missedBlockLen {
+			// we need to resize the missed block array AND update the signing info accordingly
+			switch {
+			case missedBlockLen < window:
+				// missed block array too short, lets expand it
+				indexOffset := writeInfo.SigningInfo.IndexOffset % window
+				relativeIndexOffset := writeInfo.SigningInfo.IndexOffset % missedBlockLen
+				newMissedBlocks := make([]bool, window)
+				indexOffsetCounter := indexOffset + window
+				// traverse backwards through the old array (smaller), and copy it to the relevant index offsets based on the new window and updated relative index offset with the new window size
+				for i := relativeIndexOffset + missedBlockLen; i > relativeIndexOffset; i-- {
+					missedBlockIdx := i % missedBlockLen
+					newMissedBlocks[indexOffsetCounter%window] = missedInfo.MissedBlocks[missedBlockIdx]
+					indexOffsetCounter--
+				}
+				missedInfo.MissedBlocks = newMissedBlocks
+				k.SetValidatorMissedBlocks(ctx, writeInfo.ConsAddr, missedInfo)
+			case missedBlockLen > window:
+				// missed block array too long, we need to trim
```

Comment:
how come we dont' just take the most recent values?
i.e. if we resize 5 -> 3
- before: [0..5]
- after: [3..5]
?

Reply:
we do, but we need to make sure theyre ordered properly in the resulting array such that the validators index offset will still point to the correct value AFTER the mod with the new window size